### PR TITLE
Lut rnd

### PR DESCRIPTION
**Pull request by jburel**
# What this PR does
Add support for LUT in the rendering engine.
If a LUt is set, the color on the rendering def is not used.
If a lut is not set or cannot be read, the color is used

ThumbnailBean and RenderingBean retrieve the luts using the ``scriptRepoHelper``.
Integration tests have been added/improved to test:
 * save of lut
 * read of lut. LUT readers are tested comparing the byte array of an image rendered with color and with lut.

# Testing this PR

Check that the integration tests are green.
Check that the LUT are correctly applied using insight see PR https://github.com/openmicroscopy/openmicroscopy/pull/4752
 * Use a single channel image, e.g. ``single-channel.ome.tif``
 * compare display using the same lut in both insight and imagej or fiji

# Related reading
https://trello.com/c/9iAe3Sg9/95-lut-support

### CODE REVIEW

**Comment by jburel** [comment]
Removing breaking label so @will-moore can start working on the UI and testing the workflow.

**Comment by jburel** [comment]
@pwalczysko: the issue https://github.com/openmicroscopy/openmicroscopy/pull/4752#issuecomment-234223275 should now be fixed.

**Comment by pwalczysko** [comment]
Failed checks here @jburel

**Comment by jburel** [comment]
@pwalczysko: travis issue, the last commit was not included in today's build but that's not a problem for today's review of @dominikl's PR

**Comment by jburel** [comment]
Moving to breaking so we can check the tests.

**Comment by jburel** [comment]
removed breaking label: share tests should now work, This PR should be back in the merge build tomorrow cf. https://github.com/openmicroscopy/openmicroscopy/pull/4767#issuecomment-237284813 and https://github.com/openmicroscopy/openmicroscopy/pull/4764#issuecomment-237285198
cc @dominikl @will-moore @gusferguson

**Comment by jburel** [comment]
Thumbnail: Load the lut using query service and not the scriptrepo due to issue with share and/or hibernate.This will need to be revisited when we work on the rendering engine/thumbnail service

**Comment by jburel** [comment]
@gusferguson: the "override" color issue noticed should be fixed by https://github.com/openmicroscopy/openmicroscopy/pull/4753/commits/d57c42e21cd1978c480c0ab64306613f7ec2480e
This will have to be checked on monday since this PR was not included in the build

**Comment by will-moore** [comment]
I built the server locally with your latest changes and it still doesn't look right. Here, adding TRITC channel to DAPI and GFP channels doesn't overwrite the DAPI channel, but does turn all the blue to white and the GFP green is still overwritten.

screen shot 2016-08-05 at 13 24 27

**Comment by jburel** [comment]
@will-moore: hopefully the last commit should fix the problem

**Comment by will-moore** [comment]
I just noticed that the "unionjack" LUT as shown in Fiji looks different from ours 
screen shot 2016-08-08 at 14 38 35

screen shot 2016-08-08 at 14 47 08

**Comment by jburel** [comment]
@will-moore: I got them from https://github.com/imagej/imagej-common/tree/master/src/main/resources/luts. I will check tomorrow. 
commit 54c06cc should fix the problem you reported. This PR was not included in the Today's build due to a travis failure

**Comment by jburel** [comment]
@will-moore: problem fixed, it was an issue in the lut text reader. This is the only with 4 columns (index column)

**Comment by jburel** [comment]
@will-moore: ``glasbey.lut`` and ``glow.lut`` have the same structure than ``unionjack.lut`` i.e. 4 columns
so please check that there is no issue reading those 2 luts.
The other ones are binary lut

**Comment by will-moore** [comment]
Unionjack LUT and other new ones look correct (comparing with Fiji).
Colours are now being combined correctly. 
Small issue with combining channels. E.g.
https://eel.openmicroscopy.org/merge/webclient/?show=image-25351 (user 3)
when mix "yellow hot" (channel 1) with Green (channel 2).
NB: Don't see a problem with this same image if channel 2 is switched to Blue (No blue in the "yellow hot" LUT).

screen shot 2016-08-11 at 16 52 50

**Comment by jburel** [comment]
@will-moore: the last commit should fix the "saturation" you noticed.

**Comment by will-moore** [comment]
Optimization fix looks good. Rendering nicely now - no change when mixing LUT channel with others:

screen shot 2016-08-16 at 11 05 27

**Comment by jburel** [comment]
knows issue added to https://trello.com/c/9iAe3Sg9/95-lut-support

File: components/server/src/ome/services/ThumbnailBean.java [hunk 1]
```diff
@@ -174,15 +181,17 @@
     /** Notification that the bean has just returned from passivation. */
     private transient boolean wasPassivated = false;
 
+    private final ScriptRepoHelper helper;
     /** default constructor */
-    public ThumbnailBean() {}
+    //public ThumbnailBean() {}
```
Comment (joshmoore) [inline components/server/src/ome/services/ThumbnailBean.java]: Delete?

**Comment by joshmoore** [comment]
Sorry, for coming to this late, but a few thoughts perhaps for follow-on PRs:

 * Looks like these are the old copyright headers.
 * Having a `LutReader` contain a `BasicLutReader` seems slightly confusing but isn't critical since they are package-scoped.
 * Some of these classes may exist at the B-F level. In general, it might be worth refactoring down to the OME-Files level. cc: @sbesson 

Also: especially for larger PRs like this, I'd have suggested splitting out the whitespace etc. changes.

**Comment by jburel** [comment]
I still need to work on the thumbnail bean
This code is very old. (sorry to whitespaces /tab)
 I will do some of the changes in follow-up.
I have already restarted some refactoring in the "codomain PR".

**Comment by joshmoore** [comment]
After a brief chat, @jburel will push (force?) changes for the style issues here. Refactoring will happen as part of follow-ons.

**Comment by jburel** [comment]
Following discussions and avoid possible rebasing issues, I have:

 * Remove clean up commits
 * Fix header (my eclipse template was out of date!)
 * Refactor lut readers code.
 * Open a new PR for the addition of extra binary luts i.e. https://github.com/openmicroscopy/openmicroscopy/pull/4795

If more is required, I will continue in the "codomain PR" cf. https://trello.com/c/9iAe3Sg9/95-lut-support

**Comment by jburel** [comment]
https://ci.openmicroscopy.org/job/OMERO-DEV-merge-build/ is now green. Javadoc fix
This PR is now ready to go

**Comment by joshmoore** [comment]
Thanks, @jburel. Will take a look at the next PR.

**Pull request merged**